src/core/linalg/src/dense/4C_linalg_tensor_einstein.hpp

```diff
@@ -0,0 +1,628 @@
+          return this_einstein_index.name == einstein_index.name;
+        }) > 2; })>>;
+
+
+    template <auto arr>
+    constexpr std::array contraction_einstein_indices = unique_einstein_indices<
+        filter_array<arr, decltype([](auto einstein_index) { return std::ranges::count_if(arr, [einstein_index](auto this_einstein_index){
+          return this_einstein_index.name == einstein_index.name;
+        }) == 2; })>>;
+
+    template <std::array einstein_indices>
+    constexpr bool valid_contraction_einstein_indices = []() consteval
+    {
+      for (std::size_t i = 0; i < einstein_indices.size(); ++i)
+      {
+        auto found_item = *std::ranges::find_if(
+            einstein_indices, [i](auto item) { return item.name == einstein_indices[i].name; });
+        if (found_item.size != einstein_indices[i].size)
+        {
+          return false;
+        }
+      }
+      return true;
+    }();
+
+    template <auto arr>
+    constexpr std::array dyadic_einstein_indices = []() consteval
+    {
+      std::array dyadic_einstein_indices = unique_einstein_indices<
+          filter_array<arr, decltype([](auto einstein_index) { return std::ranges::count_if(arr, [einstein_index](auto this_einstein_index){
+          return this_einstein_index.name == einstein_index.name;
+        }) == 1; })>>;
+
+      std::ranges::sort(dyadic_einstein_indices, [](auto a, auto b) { return a.name < b.name; });
+      return dyadic_einstein_indices;
+    }();
+
+    template <std::array shape, typename T, typename IndexSequence>
+    struct TensorTypeDeducer;
+
+    template <std::array shape, typename T, std::size_t... i>
+    struct TensorTypeDeducer<shape, T, std::integer_sequence<std::size_t, i...>>
+    {
+      using type = Tensor<T, shape[i]...>;
+    };
+
+    template <typename T, std::array shape>
+    using TensorTypeFromArray =
+        typename TensorTypeDeducer<shape, T, std::make_index_sequence<shape.size()>>::type;
+
+
+    /*!
+     * @brief The flat index of a component for a tensor with a compression type @p
+     * TensorCompression
+     */
+    template <typename TensorCompression, std::array index>
+    constexpr std::size_t flat_index = []() consteval
+    {
+      return [&]<std::size_t... i>(std::integer_sequence<std::size_t, i...> seq)
+      {
+        return TensorCompression::template flatten_index<Internal::TensorBoundCheck::no_check>(
+            i...);
+      }(IntegerSequenceFromArray<index>());
+    }();
+
+    /*!
+     * @brief Returns the tensor indices of a component for a tensor in an Einstein summation.
+     *
+     * Take the matrix-matrix product as an example:
+     *
+     * A_ij = B_ik C_kj.
+     *
+     * The contraction indices are [k], the dyadic indices are [i, j].
+     *
+     * Let's say, we want to evaluate the above equation for i=1, j=2, k=3, the given parameters
```

Comment:
I think my confusion might stem from `i=1, j=2, k=3`. What does this mean? I don't know how to interpret `k=3` since this is a contracted index. I'll drop by in your office tomorrow and you can enlighten me ;)